dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/uploader/BatchUploader.java

```
@@ -260,28 +289,67 @@ private boolean canEnqueueMoreRequests() {
     return client.dispatcher().queuedCallsCount() < MAX_ENQUEUED_REQUESTS;
   }
 
+  private static void enqueueCall(
+      OkHttpClient client,
+      Request request,
+      Callback responseCallback,
+      RetryPolicy retryPolicy,
+      int failureCount,
+      Phaser inflightRequests) {
+    Call call = client.newCall(request);
+    retryPolicy.failures.put(call, failureCount);
+    call.enqueue(responseCallback);
+    inflightRequests.register();
+  }
+
   private static final class ResponseCallback implements Callback {
 
     private final RatelimitedLogger ratelimitedLogger;
     private final Phaser inflightRequests;
+    private final OkHttpClient client;
+    private final RetryPolicy retryPolicy;
 
-    public ResponseCallback(final RatelimitedLogger ratelimitedLogger, Phaser inflightRequests) {
+    public ResponseCallback(
+        final RatelimitedLogger ratelimitedLogger,
+        Phaser inflightRequests,
+        OkHttpClient client,
+        RetryPolicy retryPolicy) {
       this.ratelimitedLogger = ratelimitedLogger;
       this.inflightRequests = inflightRequests;
+      this.client = client;
+      this.retryPolicy = retryPolicy;
     }
 
     @Override
-    public void onFailure(final Call call, final IOException e) {
+    public void onFailure(Call call, IOException e) {
       inflightRequests.arriveAndDeregister();
       ratelimitedLogger.warn("Failed to upload batch to {}", call.request().url(), e);
+      handleRetry(call, retryPolicy.maxNetworkFailures);
+    }
+
+    private void handleRetry(Call call, int maxFailures) {
+      Integer failure = retryPolicy.failures.remove(call);
+      if (failure != null) {
+        int failureCount = failure + 1;
+        if (failureCount <= maxFailures) {
+          LOGGER.debug("Retrying upload {}/{}", failureCount, maxFailures);
+          enqueueCall(client, call.request(), this, retryPolicy, failureCount, inflightRequests);
```

Don't want to over-complicate things, but maybe we should add a backoff? If the requests fail fast, all retries might be done in a very short time interval.

---

I see your point. The problem is how I deal with this backoff? where I put the new request on hold, who is responsible for enqueue again after x time?
also, we are keeping for a longest period large buffers (serialized symbols up to 50MB)

---

Well, we would have to keep them in a collection on the class level I guess. We could either have a scheduled thread that re-enqueues the requests or store a scheduled thread for each retry that has a given initial delay. As I said, it's making things a bit more complicated, but the more I think about it the less I like retrying immediately.